libs/postgres_versioninfo/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+
+impl PgVersionId {
+    pub const UNKNOWN: PgVersionId = PgVersionId(0);
+
+    pub fn from_full_pg_version(version: u32) -> PgVersionId {
+        match version {
+            0 => PgVersionId(version), // unknown version
+            140000..180000 => PgVersionId(version),
+            _ => panic!("Invalid full PostgreSQL version ID {version}"),
+        }
+    }
+}
+
+impl Display for PgVersionId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        u32::fmt(&self.0, f)
+    }
+}
+
+impl Serialize for PgVersionId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        u32::serialize(&self.0, serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for PgVersionId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        u32::deserialize(deserializer).map(PgVersionId)
+    }
+
+    fn deserialize_in_place<D>(deserializer: D, place: &mut Self) -> Result<(), D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        u32::deserialize_in_place(deserializer, &mut place.0)
+    }
+}
+
+impl PgMajorVersion {
+    /// Get the numerical representation of the represented Major Version
+    pub const fn major_version_num(&self) -> u32 {
+        match self {
+            PgMajorVersion::PG14 => 14,
+            PgMajorVersion::PG15 => 15,
+            PgMajorVersion::PG16 => 16,
+            PgMajorVersion::PG17 => 17,
+        }
+    }
+
+    /// Get the contents of this version's PG_VERSION file.
+    ///
+    /// The PG_VERSION file is used to determine the PostgreSQL version that currently
+    /// owns the data in a PostgreSQL data directory.
+    pub fn versionfile_string(&self) -> String {
+        match self {
+            PgMajorVersion::PG17 => "17\x0A".to_string(),
+            PgMajorVersion::PG16 => "16\x0A".to_string(),
+            PgMajorVersion::PG15 => "15".to_string(),
+            PgMajorVersion::PG14 => "14".to_string(),
+        }
+    }
+
+    /// Get the v{version} string of this major PostgreSQL version.
+    ///
+    /// Because this was hand-coded in various places, this was moved into a shared
+    /// implementation.
+    pub fn v_str(&self) -> String {
+        match self {
+            PgMajorVersion::PG17 => "v17".to_string(),
+            PgMajorVersion::PG16 => "v16".to_string(),
+            PgMajorVersion::PG15 => "v15".to_string(),
+            PgMajorVersion::PG14 => "v14".to_string(),
+        }
+    }
+
+    /// All currently supported major versions of PostgreSQL.
+    pub const ALL: [PgMajorVersion; 4] = [
+        PgMajorVersion::PG14,
+        PgMajorVersion::PG15,
+        PgMajorVersion::PG16,
+        PgMajorVersion::PG17,
+    ];
+}
+
+impl Display for PgMajorVersion {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PgMajorVersion::PG14 => {
+                write!(f, "PgMajorVersion::PG14")
```

Comment:
You could just implement this in terms of the Debug trait. Don't love that Display and Debug produce the same string though. I probably would have just done: `PGX`.

Also this could again be implemented in terms of `major_version_num()`.

Reply:
Again, I prefer <&'static str> over `format!()` if and where possible, especially when the resulting code is not that much more difficult to read. Formatting is IMO a waste of resources on these small strings.